src/Microsoft.AspNetCore.WebHooks.Receivers.Slack/SlackCommand.cs

```diff
@@ -32,171 +29,365 @@ public static class SlackCommand
+                return new KeyValuePair<StringSegment, StringSegment>(StringSegment.Empty, StringSegment.Empty);
             }
 
-            var result = new KeyValuePair<string, string>(
-                values[0].Value,
-                values.Length > 1 ? values[1].Value : string.Empty);
+            var action = enumerator.Current;
+            var value = enumerator.MoveNext() ? enumerator.Current : StringSegment.Empty;
 
-            return result;
+            return new KeyValuePair<StringSegment, StringSegment>(action, value);
         }
 
-        // ??? Should we continue to use NameValueCollection here and as the base for ParameterCollection?
         /// <summary>
-        /// Parses the 'text' of a slash command or 'subtext' of an outgoing WebHook of the form
+        /// Parses the parameters a slash command or of an outgoing WebHook of the form
         /// '<c>action param1; param2=value2; param3=value 3; param4="quoted value4"; ...</c>'.
         /// Parameter values containing semi-colons can either escape the semi-colon using a backslash character,
         /// i.e '\;', or quote the value using single quotes or double quotes.
         /// </summary>
         /// <example>
         /// An example of an outgoing WebHook or slash command using this format is
         /// <c>/appointment add title=doctor visit; time=Feb 3 2016 2 PM; location=Children's Hospital</c>
-        /// where '/appointment' is the trigger word or slash command, 'add' is the action and title, time, and location
-        /// are parameters.
+        /// where '/appointment' is the trigger word or slash command, 'add' is the action and title, time, and
+        /// location are parameters. In this case, caller would pass
+        /// <c>title=doctor visit; time=Feb 3 2016 2 PM; location=Children's Hospital</c> to this method.
         /// </example>
-        /// <param name="text">The 'text' of a slash command or 'subtext' of an outgoing WebHook.</param>
-        public static KeyValuePair<string, NameValueCollection> ParseActionWithParameters(string text)
+        /// <param name="text">
+        /// The parameter portion of a slash command or an outgoing WebHook. Often the
+        /// <see cref="KeyValuePair{StringSegment, StringSegment}.Value"/> from what
+        /// <see cref="ParseActionWithValue"/> returned.
+        /// </param>
+        /// <returns>
+        /// A <see cref="ValueTuple{T1, T2}"/> in which one property is always <see langword="null"/>. When successful,
+        /// <c>Error</c> is <see langword="null"/> and <c>Parameters</c> an <see cref="IDictionary{TKey, TValue}"/>
+        /// mapping zero or more parameter names to zero or more values. Otherwise, <c>Parameters</c> is
+        /// <see langword="null"/> and <c>Error</c> a <see cref="string"/> describing the specific problem.
+        /// </returns>
+        public static (IDictionary<StringSegment, IList<StringSegment>> Parameters, string Error) ParseParameters(
```

Comment:
I really like the whole `(parsed value, error)` pattern with tuples. Stealing this for later use 👍